### PR TITLE
Modernize build system using PEP621 semantics

### PR DESCRIPTION
**Pull request by zoj613**
This commit replaces setup.cfg+setup.py with pyproject.toml and setuptools. It also removes versioneer with setuptools-scm for automatic package versioning during development and release.

**Thank you for opening a PR!**

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [ ] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [x] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [x] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [x] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.  If your commit description starts with "Fix...", then you're probably making this mistake.
+ [ ] There are tests covering the changes introduced in the PR.

Don't worry, your PR doesn't need to be in perfect order to submit it.  As development progresses and/or reviewers request changes, you can always [rewrite the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) of your feature/PR branches.

If your PR is an ongoing effort and you would like to involve us in the process, simply make it a [draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).

### CODE REVIEW

**Comment by zoj613** [comment]
This is the resultant source distribution tree:

**Comment by zoj613** [comment]
Test failures are caused by #96

**Comment by maresb** [comment]
Hey, this looks good at first glance. I'm headed off to bed. I can take a closer look tomorrow.

Be warned that there are some really weird things that can go wrong here. What I'd do to check thoroughly is to build both the sdist and the wheel from before and after. Then do a directory comparison (I use Meld), and check that the before and after versions are equivalent. It's a good amount of work, but it's actually necessary if you want to be thorough.

**Review by brandonwillard — COMMENTED** [review]
This looks great.  After @maresb confirms the build, let's merge it.

**Comment by zoj613** [comment]
@brandonwillard black pre-commit hook is failing the tests due to lines being too long. What do we do about these? Increase the line limit or follow black's suggestions?

**Comment by brandonwillard** [comment]
> @brandonwillard black pre-commit hook is failing the tests due to lines being too long. What do we do about these? Increase the line limit or follow black's suggestions?

It looks like some of the config options in the old `setup.cfg` aren't being picked up in the new setup.  Regarding `black`, that could be due to a new version and/or the aforementioned.

**Comment by zoj613** [comment]
> > @brandonwillard black pre-commit hook is failing the tests due to lines being too long. What do we do about these? Increase the line limit or follow black's suggestions?
> 
> It looks like some of the config options in the old `setup.cfg` aren't being picked up in the new setup. Regarding `black`, that could be due to a new version and/or the aforementioned.

I think this is due to the fact that there aren't any hook settings for black specified in the pre-commit config and there is no black tool section in the `pyproject.toml` specifying `black` specific settings to override black's default settings. So we probably need to decide on sensible settings to add in pyproject.toml

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,131 @@
+[build-system]
+requires = ["wheel", "setuptools>=61.2", "setuptools-scm[toml]>=6.2"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "aehmc"
+authors = [
+    {name = "aesara-devs", email = "aesara.devs@gmail.com"}
```
Comment (maresb) [inline pyproject.toml:8]: Minor aesthetic point: I think we can change the name to "Aesara developers" both here and in the Aesara repo.

File: MANIFEST.in [hunk 2]
```diff
@@ -1,2 +1,8 @@
-include versioneer.py
-include aehmc/_version.py
+prune .github
+prune tests
```
Comment (maresb) [inline MANIFEST.in:2]: You may wish to consider leaving in `tests`. Although they aren't required for runtime execution, they don't take up much space and it is useful for package maintainers to have them. For example, see https://github.com/pypa/setuptools_scm/issues/806#issuecomment-1442254345

**Review by maresb — COMMENTED** [review]
The pre-commit errors are arising from flake8 not black. See the corresponding comment below about how to fix it. Thanks @zoj613!!!

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+force_grid_wrap = 0
+use_parentheses = true
+ensure_newline_before_comments = true
+line_length = 88
+
+[tool.flake8]
+max-line-length = "88"
+select = "C,E,F,W"
+ignore = "E203,E231,E501,E741,W503,W504,C901"
+per-file-ignores = """
```
[the remark this answers is not in the document]
Comment (zoj613) [reply pyproject.toml]: I wasn't aware, thanks for this!

**Comment by zoj613** [comment]
Thanks for the review @maresb , I've added the suggestions in [`ec01897` (#97)](https://github.com/aesara-devs/aehmc/pull/97/commits/ec01897cef1f0dfb29bb9fe6106c660fc594dab6), please do take another look when you can.

**Comment by zoj613** [comment]
@brandonwillard @rlouf any particular comments before merge? I'll squash the 7 commits into one If there aren't any.

**Comment by brandonwillard** [comment]
> @brandonwillard @rlouf any particular comments before merge? I'll squash the 7 commits into one If there aren't any.

Yeah, squash and merge is fine.

**Comment by zoj613** [comment]
> > @brandonwillard @rlouf any particular comments before merge? I'll squash the 7 commits into one If there aren't any.
> 
> Yeah, squash and merge is fine.

done!

**Comment by brandonwillard** [comment]
> > > @brandonwillard @rlouf any particular comments before merge? I'll squash the 7 commits into one If there aren't any.
> > 
> > 
> > Yeah, squash and merge is fine.
> 
> done! I don't have merge permissions for `aehmc` so please do the merge once the CI run is done.

One second, I'll add you to the group.

**Comment by brandonwillard** [comment]
Done; try it now.

**Comment by zoj613** [comment]
> Done; try it now.

merging is still blocked. Not sure if another setting needs to be changed.

**Comment by maresb** [comment]
Looks like a second review is required?
image

**Comment by brandonwillard** [comment]
OK, now?

**Comment by brandonwillard** [comment]
Much appreciated, @zoj613!

**Pull request merged**